[Implement] : FAB

wndudqus opened the pull request:
## Description
[implement] : FAB
1. Implemented the basic functions of the FAB
        * Can use IconButton that is a default button component of FAB by passing the icon types and callbacks.
        * Can replace the default button component of FAB by passing renderButton function
        * Can set absolute position of FAB
        * Can replace buttons by active Props

Rendering conditions:
         if(active===false)render DefaultFAB
         if(active===true)render ActiveFAB and ActionButtons

please comment about the structure of Component.

    **Animation is not implemented**
    **Need more implementation**

## Related Issues

[[FloatingActionButton] implementing floatingActionButton component](https://github.com/dooboolab/dooboo-ui/issues/63#issue-964570899)
## Tests

I did not add the test yet.
This PR needs more implementation.

## Checklist

- [X] I read the [Contributor Guide](https://github.com/dooboolab/dooboo-ui/blob/master/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [X] Run `yarn test` or `yarn test -u` if you need to update snapshot.
- [X] Run `yarn lint`
- [X] I am willing to follow-up on review comments in a timely manner.

yujong-lee commented:
I slightly look into it. I think maybe it will be easier for me to understand if you provide some example code and output.
But It was totally fine.

Ok. As I understand,  `FABList` is rendered only when `Active`. And `DefaultFAB`/`ActiveFAB` is chosen by `Active` state.
Also, `ActiveFAB`, `DefaultFAB`, and `FABItem` can have `render function` or it is rendered as default form.

I think overall structure is fine, but `naming` can be better.

1. I think `DefaultFAB` is the most confusing part.  `Default <--> Active` should be replaced with `Inactive <--> Active` or `Expanded <--> Collapsed`. And the word `Default` should be used for `without render function` situation.
2. `ActiveFAB` and  `FABItem` has `Parent-Children` relation, but it is not revealed with name.

- - - 
BTW,  this might be silly mistake of mine, but I'm seeing icon never changes.
I'm always seeing same icon whenever I change iconName string.

I'm using like this.
```js
<FAB
  active={true}
  ActiveFAB={{icon: 'cross-light', onPress: () => null}}
  DefaultFAB={{icon: 'cross-light', onPress: () => null}}
  FABList={[
    {icon: 'chevron-down-light', onPress: () => null},
    {icon: 'chevron-down-light', onPress: () => null},
    {icon: 'chevron-down-light', onPress: () => null},
  ]}
/>
```

yujong-lee commented on main/FloatingActionButtonWrapper/index.tsx:
```diff
@@ -0,0 +1,120 @@
+            ),
+          )}
+      </Animated.View>
+      <FABItemWrapperView>{active ? defaultFAB : activeFAB}</FABItemWrapperView>
```
Maybe you mean `{active ? activeFAB : defaultFAB}`

wndudqus replied:
Oh yes, it is thank you for pointing it out!

wndudqus replied:
I strongly agree with your opinion about naming.
I will change the naming to more self-descriptively.

And you should use like this. 
( the active button and default button was same in your code )

```TS
<FAB
  active={true}
  ActiveFAB={{icon: 'cross-light', onPress: () => null}}
  DefaultFAB={{icon: 'add-light', onPress: () => null}}
  FABList={[
    {icon: 'chevron-down-light', onPress: () => null},
    {icon: 'chevron-down-light', onPress: () => null},
    {icon: 'chevron-down-light', onPress: () => null},
  ]}
/>
```
I talked about naming and how to simplify the props of this component with @hyochan 
And I decided to not use DefaultFAB and ActiveFAB instead, I will support renderMainFAB for customizing the DefaultButton 

It will look like this

```TS
 <FAB
          theme={theme}
          isActive={isActive}
          FABList={[{id: 'search', icon: 'home-light'}]}
          onPressFABItem={(item) => {
            switch (item.id) {
              case 'mainDefault':
                setFabActive(true);
                break;

              case 'mainActive':
                setFabActive(false);
                break;

              default:
                break;
            }
          }}
        />
```

yujong-lee commented:
It's hard to communicate many changes at once with `review`(and `codecov` warning is annoying 😄 ), so I pasted code below.
This is just a suggestion. Feel free to point out anything. :)

1. I can't be sure this refactor guarantee same functionality as before because there's no test code. But I think It is sufficient for suggestion.
2.  `theme` is not provided to `defaultSubRenderer` in code below. It should be changed. maybe `useTheme` is needed?

```jsx
const defaultRootRenderer = ({icon, onPress}: FABItem): React.ReactElement => (
  <IconButton icon={<StyledIcon size={24} name={icon} />} onPress={onPress} />
);

const defaultSubRenderer = (
  {icon, onPress}: FABItem,
  _index: number,
): React.ReactElement => (
  <IconButton
    icon={<StyledIcon theme={theme} size={24} name={icon} />}
    onPress={onPress}
  />
);

const FloatingActionButtons: FC<
  FloatingActionButtonsWrapperProps & {theme: DoobooTheme}
> = ({
  theme,
  active,
  inactiveRootItem,
  activeRootItem,
  subItems,
  renderInactiveRootItem = defaultRootRenderer,
  renderActiveRootItem = defaultRootRenderer,
  renderSubItem = defaultSubRenderer,
  zIndex = 0,
  top = 'auto',
  bottom = 'auto',
  left = 'auto',
  right = 'auto',
}) => {
  return (
    <FABContainer
      style={{
        top,
        bottom,
        left,
        right,
        zIndex,
      }}>
      <Animated.View>
        {active &&
          subItems.map((item, idx) => (
            <ItemContainer key={item.icon + idx}>
              {renderSubItem(item, idx)}
            </ItemContainer>
          ))}
      </Animated.View>
      <ItemContainer>
        {active
          ? renderInactiveRootItem(inactiveRootItem)
          : renderActiveRootItem(activeRootItem)}
      </ItemContainer>
    </FABContainer>
  );
};

export const FAB = withTheme(FloatingActionButtons);
```

yujong-lee commented on main/FloatingActionButtonWrapper/index.tsx:
```diff
@@ -0,0 +1,120 @@
+import {IconButton} from '../IconButton';
+import styled from '@emotion/native';
+
+const AbsolutePositionWrapperView = styled.View`
```
I think `AbsolutePositionWrapperView` is too verbose. I'd rather use code below.
```js
<View style={{position: 'absolute'}}>
</View>
```

Considering further changes and abstraction, name it to `~Container` might be better,

wndudqus replied:
I think it is good thanks.  👍 
I thought I'd use more styles when I first added it.( But I didn't )

wndudqus replied:
but I think we should support default components such as <IconButton> with conditionals.
Because it makes it much easier to use components. It will improve developers' experience in use.

yujong-lee replied:
@wndudqus  For sure. My code above support default components, but I remove conditional using default value which language provide.

```
// code from props destructing above
renderInactiveRootItem = defaultRootRenderer,
renderActiveRootItem = defaultRootRenderer,
renderSubItem = defaultSubRenderer,
```

Note that I set default render function that I defined above. User can provide render function, or omit it.

yujong-lee commented on main/FloatingActionButtonWrapper/index.tsx:
```diff
@@ -0,0 +1,120 @@
+  renderActiveFAB?: (item: FABItem) => React.ReactElement;
+  renderFABListItem?: (item: FABItem, idx: number) => React.ReactElement;
+  zIndex?: number;
+  top?: number;
+  bottom?: number;
+  left?: number;
+  right?: number;
```
These prop should be wrapped with `style` object.
@hyochan wants components in `dooboo-ui` has `style` and `styles` props.
See other component(`Button` is good example) for more detail. 

Or maybe @hyochan can explain this better than me.

wndudqus replied:
thanks, I will change It to StyleProp<ViewStyle> and set the default value

hyochan reviewed (commented):
Looks like @yujong-lee covered most of my review already which I've communicated earlier. Thanks a lot! You are such a quick learner 💯

hyochan commented on main/FloatingActionButtonWrapper/index.tsx:
```diff
@@ -0,0 +1,120 @@
+
+interface FloatingActionButtonsWrapperProps {
+  active: boolean;
+  DefaultFAB: FABItem;
```
I think this better naming would be `MainFab`. Default sounds like something before customized.

hyochan commented on main/FloatingActionButtonWrapper/index.tsx:
```diff
@@ -0,0 +1,120 @@
+  DefaultFAB: FABItem;
+  ActiveFAB: FABItem;
+  FABList: FABItem[];
+  renderDefaultFAB?: (item: FABItem) => React.ReactElement;
```
and renderMainFAB

hyochan commented on main/FloatingActionButtonWrapper/index.tsx:
```diff
@@ -0,0 +1,120 @@
+`;
+
+export type FABItem = {
+  icon: IconName;
```
Why restricting to `icon` name? `ReactElement` will be much flexible.

hyochan reviewed (commented):
I think the spec would be simplified to below. How do you think 🤔 ?
```
  <FAB
    style={{ background: 'red' }}
    data={[ 'main', 'settings', 'addItem' ]}
    icon="plus"
    isActive={isActive}
    onPress={() => setIsActive(!isActive)}
    onPressItem={(data) => console.log('pressed', data)}
    renderFAB={() => <View/>}
    renderFABItem={() => <View/>}
  />
```

hyochan commented:
Also the generic can be extended like below
```
interface FABItemProp {
  icon: string;
  name: string;
}

type Props<T extends FABItemProp> = {
```

wndudqus commented:
@hyochan  thanks 
It is very good idea to use generic extended the FABItem.
It will give more flexibility to  FAB component

yujong-lee commented:
> but I think we should support default components such as with conditionals.
> Because it makes it much easier to use components. It will improve developers' experience in use.

@wndudqus  For sure. My code above support default components, but I remove conditional using `default value feature` which language provides.

```
// code from props destructing above
renderInactiveRootItem = defaultRootRenderer,
renderActiveRootItem = defaultRootRenderer,
renderSubItem = defaultSubRenderer,
```

Note that I set default render function that I defined above. User can provide render function, or omit it.

Is there's anything I misunderstood, please let me know.

wndudqus commented:
@yujong-lee  
Oh, I see... I think this approach seems to simplify the code. 
How do you think? @hyochan

hyochan commented:
> the

Yes for the default component~ You can try @yujong-lee 's proposal 👍

hyochan commented:
> @yujong-lee
> Oh, I see... I think this approach seems to simplify the code.
> How do you think? @hyochan

Yes for the default component~ You can try @yujong-lee 's proposal 👍

hyochan commented on main/FloatingActionButtonWrapper/index.tsx:
```diff
@@ -0,0 +1,120 @@
+
+export interface FABProps<ITEM extends FABItem> {
+  ButtonList: ITEM[];
+  onPressListItem: (item: ITEM | FABItem) => void;
```
I think `ITEM | FABItem` => `FabItem`

hyochan replied:
Also I think this can be `optional`

hyochan commented on main/__tests__/FAB.test.tsx:
```diff
@@ -0,0 +1,32 @@
+
+    const {getByTestId} = render(
+      Component({
+        ButtonList: [item1],
```
`ButtonList` => `fabItems`?

hyochan commented on stories/dooboo-ui/FloatingActionButton/FABStory.tsx:
```diff
@@ -0,0 +1,43 @@
+      <SafeAreaView style={{width: '100%', height: '100%'}}>
+        <FAB
+          size="medium"
+          theme={theme}
```
Removing `theme` props here and wrap with `ThemeProvider` would be a better example.

hyochan commented on stories/dooboo-ui/FloatingActionButton/FABStory.tsx:
```diff
@@ -0,0 +1,43 @@
+            {id: 'search', icon: 'home-light'},
+            {id: 'like', icon: 'like-light'},
+          ]}
+          onPressListItem={(item) => {
```
```
onPressListItem={({item, index}) => {
```
Support `index`?

hyochan commented on main/index.ts:
```diff
@@ -14,3 +14,4 @@ export * from './Accordion';
 export * from './StatusBarBrightness';
 export * from './SwitchToggle';
 export * from './theme';
+export * from './FloatingActionButtonWrapper';
```
`./FAB`?

hyochan commented on main/FloatingActionButtonWrapper/index.tsx:
```diff
@@ -0,0 +1,120 @@
+`;
+
+const FABItemWrapperView = styled.View`
+  position: relative;
```
I think this is not necessary.

hyochan commented on main/FloatingActionButtonWrapper/index.tsx:
```diff
@@ -0,0 +1,120 @@
+}
+
+export interface FABProps<ITEM extends FABItem> {
+  ButtonList: ITEM[];
```
`Item`?

hyochan commented on main/FloatingActionButtonWrapper/index.tsx:
```diff
@@ -0,0 +1,120 @@
+export interface FABProps<ITEM extends FABItem> {
+  ButtonList: ITEM[];
+  onPressListItem: (item: ITEM | FABItem) => void;
+  renderMainFAB?: (isActive: boolean, onPressFAB: () => void) => ReactElement;
```
`renderFAB`?

hyochan commented on main/FloatingActionButtonWrapper/index.tsx:
```diff
@@ -0,0 +1,120 @@
+  ButtonList: ITEM[];
+  onPressListItem: (item: ITEM | FABItem) => void;
+  renderMainFAB?: (isActive: boolean, onPressFAB: () => void) => ReactElement;
+  renderFABListItem?: (item: ITEM, idx: number) => ReactElement;
```
`renderFabItem`?

hyochan commented on main/FloatingActionButtonWrapper/index.tsx:
```diff
@@ -0,0 +1,120 @@
+    <View
+      style={
+        style || {
+          right: '5%',
+          bottom: '5%',
```
How about giving default `number`? The percentage usually cause unexpected behavior

hyochan replied:
```
style={[
   {position: 'absolute', zIndex: 999, right: 40, bottom: 40},
   style,
]}
```

hyochan reviewed (changes requested):
Great work~! I've just pointed small fixes on `naming`.

Also, I'd like you to check if it works in `web` and `android`.

hyochan commented:
![1uuezORGeM](https://user-images.githubusercontent.com/27461460/131250413-66de29f1-b85a-4d40-b567-7b832a46722d.gif)

When `FAB` first renders, it doesn't hide the children. It should not run the animation when it shouldn't and also should follow `isActive` prop.

When looking in your code I see that you are showing children with `opacity` animation.
<img width="936" alt="Screen Shot 2021-08-29 at 9 33 28 PM" src="https://user-images.githubusercontent.com/27461460/131250479-b247840a-c64e-4b92-9333-8f90179d257c.png">

I have some issues with the above design.

1. The `FABItem` should appear from the main `FAB` button but they are just appearing from their target positions. I think we need to change that.
2. As told previously, it should behave with actual prop value.

hyochan commented:
Additionally, the custom `render` function would not want to receive the callback for the `press` event. It'd only want to provide a different UI element.

```
{FABList.map((item, idx) => {
  return (
    <TouchableOpacity onPress={onPressFABItem?.(item)}>
      {renderFABListItem ? (
        renderFABListItem(item, idx)
      ) : (
        <StyledIcon theme={theme} size={24} name={item.icon} />
      )}
    </TouchableOpacity>
  );
})}
```

wndudqus commented:
@hyochan @yujong-lee 

I think I covered most of your reviews 

 1. change Props naming
 2. changed behavior
             Previously, only the transparency of the button was adjusted depending on the state but now the actual position is adjusted.
    3. Change state management.
            Previously, the active was managed as a "state", modified to manage it as a "props".
    4.  path name changed
            FloatingActionButtonWrapper> FAB
   5. Bug that state was not synced with UI has fixed

https://user-images.githubusercontent.com/35381940/133622429-09a1282c-e65c-4006-84d8-91daa1f9c508.mp4

yujong-lee commented on main/FAB/index.tsx:
```diff
@@ -0,0 +1,127 @@
+  const spinValue = useRef(new Animated.Value(0));
+  const positionValue = useRef(new Animated.Value(0));
+
+  useLayoutEffect(() => {
+    Animated.timing(spinValue.current, {
+      toValue: isActive ? 1 : 0,
+      duration: 300,
+      easing: Easing.linear,
+      useNativeDriver: true,
+    }).start();
+  }, [isActive]);
+
+  useLayoutEffect(() => {
+    Animated.timing(positionValue.current, {
+      toValue: isActive ? 1 : 0,
+      duration: 300,
+      easing: Easing.linear,
+      useNativeDriver: true,
+    }).start();
+  }, [isActive]);
```
I suggest you using `Animated.parallel`.

```js
useLayoutEffect(() => {
  const config = {
    toValue: isActive ? 1 : 0,
    duration: 300,
    easing: Easing.linear,
    useNativeDriver: true,
  };

  Animated.parallel([
    Animated.timing(spinValue.current, config),
    Animated.timing(positionValue.current, config),
  ]).start();
}, [isActive]);
```

yujong-lee commented on main/FAB/index.tsx:
```diff
@@ -0,0 +1,127 @@
+  onPressFabItem,
+  renderFAB,
+  renderFabItem,
+  size = 'large',
```
Although they might have same size in general, I think `FABItem` is responsible for setting size, not `FloatingActionButtons` component.

yujong-lee commented on main/FAB/index.tsx:
```diff
@@ -0,0 +1,127 @@
+
+export const StyledIcon = styled(Icon)`
+  color: ${({theme}) => theme.textContrast};
+`;
```
Need line break here. 😄

yujong-lee commented on main/FAB/index.tsx:
```diff
@@ -0,0 +1,127 @@
+    }).start();
+  }, [isActive]);
+
+  const spin = spinValue.current.interpolate({
+    inputRange: [0, 1],
+    outputRange: ['0deg', '45deg'],
+  });
```
I think It would've been difficult to name it because `spinValue` was already defined.

I usually take this difficulty as a sign of changing the structure. 
I suggest you to inline this in `transform : rotate`.

yujong-lee commented on main/FAB/index.tsx:
```diff
@@ -0,0 +1,127 @@
+          <IconButton
+            testID={'main_fab'}
+            size={size}
+            icon={<StyledIcon size={24} name="add-light" />}
```
Currently, user can set size of `IconButton`(small, medium, large), but not `Icon`(24 fixed).  I think we should set size of icon proportion to `IconButton`, or user can set it manually.

yujong-lee commented on main/FAB/index.tsx:
```diff
@@ -0,0 +1,127 @@
+    outputRange: ['0deg', '45deg'],
+  });
+
+  const offsetAnimationList = useMemo(
```
I suggest you to rename it to something like `offsets`. Because it is not a list of animations.

yujong-lee commented on main/FAB/index.tsx:
```diff
@@ -0,0 +1,127 @@
+  const offsetAnimationList = useMemo(
+    () =>
+      fabItems?.map((_, idx) => {
+        return positionValue.current.interpolate({
```
I personally think omitting `return` is better for readability. But it is just a suggestion.

yujong-lee reviewed (changes requested):
I was impressed by your clean code. 👍 

I just left few more suggestions you can try!

yujong-lee commented on main/FAB/index.tsx:
```diff
@@ -0,0 +1,127 @@
+
+export interface FABProps<Item extends FABItem> {
+  isActive: boolean;
+  fabItems: Item[];
```
I think It should be `FABItems`.

yujong-lee commented on main/FAB/index.tsx:
```diff
@@ -0,0 +1,127 @@
+  onPressFAB: () => void;
+  onPressFabItem: (item?: Item) => void;
+  renderFAB?: () => ReactElement;
+  renderFabItem?: (item: Item, idx: number) => ReactElement;
```
I think it should be `renderFABItem`.

yujong-lee commented:
There's still some parts that is not reflected in code review. For example, [renderProp](https://github.com/dooboolab/dooboo-ui/pull/67#discussion_r710241293), or `styles`.(`buttonWrapperStyle` is better replaced with `styles`.)

**If some reviews are not acceptable, feel free to tell me and have a discussion. It should not be one-sided.** 

But since this pull request is old and big, let's take care of this in later PR.
Thank you for great work on this!

wndudqus commented:
@yujong-lee   oh I left some comment about your review like renderProp, but I forgot to submit the review so.. my mistake sorry
